src/usr.sbin/pppd/pppstats/pppstats.8
.\"	@(#) $Header$ (LBL)
.Dd November 15, 1994
.Dt PPPSTATS 8
.Sh NAME
.Nm pppstats
.Nd report statistics for the specified PPP interface
.Sh SYNOPSIS
.Nm pppstats
.Op Fl v
.Op Fl r
.Op Fl z
.Op Fl c Ar count
.Op Fl w Ar wait
.Op Ar interface
.Sh DESCRIPTION
.Pp
The 
.Nm pppstats
utility reports PPP-related statistics at regular intervals for the
specified PPP
.Ar interface .
If the
.Ar interface
is unspecified, it will default to ppp0.
The display is split horizontally
into input and output sections containing columns of statistics
describing the properties and volume of packets received and
transmitted by the interface.
.Pp
The first report will consist of a snapshot of the interface's
statistics.  Further output, if any, will describe activity between
report intervals.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar count
Repeat the display
.Ar count
times. The first display is for the time since a reboot and each
subsequent report is for the time period since the last display.
If this option is not specified, the default repeat
.Ar count
is 1 if the
.Fl w
option is not specified, otherwise infinity.
.It Fl r
Display additional statistics summarizing the compression ratio
achieved by the packet compression algorithm in use.
.It Fl v
Display additional statistics demonstrating the efficacy of VJ header
compression and providing more explicit information on the operation
of the algorithm.
.It Fl w Ar wait
Pause
.Ar wait
seconds between each display. If no
.Ar wait
interval is specified, the default is 5 seconds.
.It Fl z
Instead of the standard display, show statistics indicating the
performance of the packet compression algorithm in use.
.El
.Pp
The following fields are printed on the input side when the
.Fl z
option is not used:
.Bl -tag -width search
.It Li IN
The total number of bytes received by this interface.
.It Li PACK
The total number of packets received by this interface.
.It Li VJCOMP
The number of compressed TCP packets received by this interface.
.It Li VJUNC
The number of uncompressed TCP packets received by this interface.
.It Li VJERR
The number of corrupted or bogus packets received by this interface.
.It Li VJTOSS
The number of VJ header-compressed TCP packets dropped on reception by
this interface.  Only reported when the
.Fl v
option is specified.
.It Li NON-VJ
The total number of non-TCP packets received by this interface. Only
reported when the
.Fl v
option is specified.
.It Li RATIO
The overall compression ratio achieved for received packets by the
packet compression scheme in use, as a number between 0.0 and 1.0,
representing the average proportionate reduction in the size of the
packets achieved by the use of packet compression (thus 0.0 indicates
that the data is incompressible).  Only reported when the
.Fl r
option is specified.
.It Li UBYTE
The total number of bytes received, after decompression of compressed
packets.  Only reported when the
.Fl r
option is specified.
.El
.Pp
The following fields are printed on the output side:
.Bl -tag -width search
.It Li OUT
The total number of bytes transmitted from this interface.
.It Li PACK
The total number of packets transmitted from this interface.
.It Li VJCOMP
The number of TCP packets transmitted from this interface with
VJ-compressed TCP headers.
.It Li VJUNC
The number of TCP packets transmitted from this interface with
VJ-uncompressed TCP headers.
.It Li NON-VJ
The total number of non-TCP packets transmitted from this interface.
.It Li VJSRCH
The number of searches for the cached header entry for a VJ header
compressed TCP packet.  Only reported when the
.Fl v
option is specified.
.It Li VJMISS
The number of failed searches for the cached header entry for a
VJ header compressed TCP packet.  Only reported when the
.Fl v
option is specified.
.It Li RATIO
The overall compression ratio achieved for transmitted packets by the
packet compression scheme in use, as a number between 0.0 and 1.0.
Only reported when the
.Fl r
option is specified.
.It Li UBYTE
The total number of bytes to be transmitted, before packet compression
is applied.  Only reported when the
.Fl v
option is specified.
.El
.Pp
When the
.Fl z
option is specified,
.Nm pppstats
instead displays the following fields, relating to the packet
compression algorithm currently in use.  If packet compression is not
in use, these fields will all display zeroes.  The fields displayed on
the input side are:
.Bl -tag -width search
.It Li COMPRESSED BYTE
The number of bytes of compressed packets received.
.It Li COMPRESSED PACK
The number of compressed packets received.
.It Li INCOMPRESSIBLE BYTE
The number of bytes of incompressible packets (that is, those which
were transmitted in uncompressed form) received.
.It Li INCOMPRESSIBLE PACK
The number of incompressible packets received.
.It Li COMP RATIO
The recent compression ratio for incoming packets, as a number between
0.0 and 1.0.
.El
.Pp
The fields displayed on the output side are:
.Bl -tag -width search
.It Li COMPRESSED BYTE
The number of bytes of compressed packets transmitted.
.It Li COMPRESSED PACK
The number of compressed packets transmitted.
.It Li INCOMPRESSIBLE BYTE
The number of bytes of incompressible packets transmitted (that is,
those which were transmitted in uncompressed form).
.It Li INCOMPRESSIBLE PACK
The number of incompressible packets transmitted.
.It Li COMP RATIO
The recent compression ratio for outgoing packets, as a number between
0.0 and 1.0.
.El
.Sh SEE ALSO
.Xr pppd 8


src/usr.sbin/pppd/pppstats/pppstats.c
/*
 * print PPP statistics:
 * 	pppstats [-v] [-r] [-z] [-c count] [-w wait] [interface]
 *
 *   -v Verbose mode for default display
 *   -r Show compression ratio in default display
 *   -z Show compression statistics instead of default display
 *
 * from the original "slstats" by Van Jacobson
 *
 * Copyright (c) 1989 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 *	Van Jacobson (van@helios.ee.lbl.gov), Dec 31, 1989:
 *	- Initial distribution.
 */

#ifndef lint
static char rcsid[] = "$Id$";
#endif

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <net/ppp_defs.h>
#include <net/if.h>
#include <net/if_ppp.h>

int	vflag, rflag, zflag;	/* select type of display */
int	interval, count;
int	infinite;
int	unit;
int	s;			/* socket file descriptor */
int	signalled;		/* set if alarm goes off "early" */
char	*progname;
char	interface[IFNAMSIZ];

void
usage()
{
	fprintf(stderr, "Usage: %s [-v|-r|-z] [-c count] [-w wait] [interface]\n",
		progname);
	exit(1);
}

/*
 * Called if an interval expires before intpr has completed a loop.
 * Sets a flag to not wait for the alarm.
 */
void
catchalarm(arg)
    int arg;
{
	signalled = 1;
}

void
get_ppp_stats(curp)
    struct ppp_stats *curp;
{
	struct ifpppstatsreq req;

	strncpy(req.ifr_name, interface, sizeof(req.ifr_name));
	if (ioctl(s, SIOCGPPPSTATS, &req) == 0) {
		*curp = req.stats;
		return;
	}
	fprintf(stderr, "%s: ", progname);
	if (errno == ENOTTY)
		errx(1, "kernel support missing");
	else
		err(1, "couldn't get PPP statistics");
}

get_ppp_cstats(csp)
    struct ppp_comp_stats *csp;
{
	struct ifpppcstatsreq creq;

	strncpy(creq.ifr_name, interface, sizeof(creq.ifr_name));
	if (ioctl(s, SIOCGPPPCSTATS, &creq) == 0) {
		*csp = creq.stats;
		return;
	}
	if (errno == ENOTTY) {
		if (zflag)
			errx(1, "no kernel compression support\n");
		warnx("no kernel compression support\n");
		rflag = 0;
	} else
		err(1, "couldn't get PPP compression stats");
}


#define V(offset)	(cur.offset - old.offset)
#define W(offset)	(ccs.offset - ocs.offset)

#define CRATE(x)	((x).unc_bytes == 0? 0.0: \
			 1.0 - ((double)((x).comp_bytes + (x).inc_bytes) \
				/ (x).unc_bytes))

/*
 * Print a running summary of interface statistics.
 * Repeat display every interval seconds, showing statistics
 * collected over that interval.  Assumes that interval is non-zero.
 * First line printed is cumulative.
 */
void
intpr()
{
	register int line = 0;
	sigset_t oldmask, mask;
	struct ppp_stats cur, old;
	struct ppp_comp_stats ccs, ocs;

	memset(&old, 0, sizeof(old));
	memset(&ocs, 0, sizeof(ocs));

	while (1) {
		get_ppp_stats(&cur);
		if (zflag || rflag)
			get_ppp_cstats(&ccs);

		(void)signal(SIGALRM, catchalarm);
		signalled = 0;
		(void)alarm(interval);
    
		if ((line % 20) == 0) {
			if (zflag) {
				printf("IN:  COMPRESSED  INCOMPRESSIBLE   COMP | ");
				printf("OUT: COMPRESSED  INCOMPRESSIBLE   COMP\n");
				printf("    BYTE   PACK     BYTE   PACK  RATIO | ");
				printf("    BYTE   PACK     BYTE   PACK  RATIO");
			} else {
				printf("%8.8s %6.6s %6.6s %6.6s %6.6s",
					"IN", "PACK", "VJCOMP", "VJUNC", "VJERR");
				if (vflag)
					printf(" %6.6s %6.6s", "VJTOSS", "NON-VJ");
				if (rflag)
					printf(" %6.6s %6.6s", "RATIO", "UBYTE");
				printf("  | %8.8s %6.6s %6.6s %6.6s %6.6s",
					"OUT", "PACK", "VJCOMP", "VJUNC", "NON-VJ");
				if (vflag)
					printf(" %6.6s %6.6s", "VJSRCH", "VJMISS");
				if(rflag)
					printf(" %6.6s %6.6s", "RATIO", "UBYTE");
			}
			putchar('\n');
		}
	
		if (zflag) {
			printf("%8u %6u %8u %6u %6.2f",
			       W(d.comp_bytes),
			       W(d.comp_packets),
			       W(d.inc_bytes),
			       W(d.inc_packets),
			       ccs.d.ratio == 0? 0.0:
			        1 - 1.0 / ccs.d.ratio * 256.0);

			printf(" | %8u %6u %8u %6u %6.2f",
			       W(c.comp_bytes),
			       W(c.comp_packets),
			       W(c.inc_bytes),
			       W(c.inc_packets),
			       ccs.c.ratio == 0? 0.0:
			        1 - 1.0 / ccs.c.ratio * 256.0);
	
		} else {

			printf("%8u %6u %6u %6u %6u",
			       V(p.ppp_ibytes),
			       V(p.ppp_ipackets),
			       V(vj.vjs_compressedin),
			       V(vj.vjs_uncompressedin),
			       V(vj.vjs_errorin));
			if (vflag)
				printf(" %6u %6u",
				       V(vj.vjs_tossed),
				       V(p.ppp_ipackets) -
				       V(vj.vjs_compressedin) -
				       V(vj.vjs_uncompressedin) -
				       V(vj.vjs_errorin));
			if (rflag)
				printf(" %6.2f %6u",
				       CRATE(ccs.d),
				       W(d.unc_bytes));
			printf("  | %8u %6u %6u %6u %6u",
			       V(p.ppp_obytes),
			       V(p.ppp_opackets),
			       V(vj.vjs_compressed),
			       V(vj.vjs_packets) - V(vj.vjs_compressed),
			       V(p.ppp_opackets) - V(vj.vjs_packets));
			if (vflag)
				printf(" %6u %6u",
				       V(vj.vjs_searches),
				       V(vj.vjs_misses));

			if (rflag)
				printf(" %6.2f %6u",
				       CRATE(ccs.c),
				       W(c.unc_bytes));
	    
		}

		putchar('\n');
		fflush(stdout);
		line++;

		count--;
		if (!infinite && !count)
			break;

		oldmask = sigblock(sigmask(SIGALRM));
		if (signalled == 0)
			sigpause(0);
		sigsetmask(oldmask);
		signalled = 0;
		(void)alarm(interval);
		old = cur;
		ocs = ccs;
	}
}

int
main(argc, argv)
	int argc;
	char *argv[];
{
	int c;
	struct ifreq ifr;

	strcpy(interface, "ppp0");
	if ((progname = strrchr(argv[0], '/')) == NULL)
		progname = argv[0];
	else
		++progname;

	while ((c = getopt(argc, argv, "vrzc:w:")) != -1) {
		switch (c) {
		case 'v':
			++vflag;
			break;
		case 'r':
			++rflag;
			break;
		case 'z':
			++zflag;
			break;
		case 'c':
			count = atoi(optarg);
			if (count <= 0)
				usage();
			break;
		case 'w':
			interval = atoi(optarg);
			if (interval <= 0)
				usage();
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (!interval && count)
		interval = 5;
	if (interval && !count)
		infinite = 1;
	if (!interval && !count)
		count = 1;

	if (argc > 1)
		usage();
	if (argc > 0)
		strncpy(interface, argv[0], sizeof(interface));

	if (sscanf(interface, "ppp%d", &unit) != 1)
		errx(1, "invalid interface '%s' specified\n", interface);
	s = socket(AF_INET, SOCK_DGRAM, 0);
	if (s < 0)
		err(1, "couldn't create IP socket");
	strcpy(ifr.ifr_name, interface);
	if (ioctl(s, SIOCGIFFLAGS, (caddr_t)&ifr) < 0)
		errx(1, "nonexistent interface '%s' specified\n", interface);

	intpr();
	exit(0);
}
